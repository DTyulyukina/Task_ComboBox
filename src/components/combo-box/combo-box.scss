$default_color: #6495ED;
$combobox_color: #696969;
$option_list_color: #D3D3D3;
$scroll_color: #C0C0C0;
$list_color: #999;
$white_color: white;

.combo-box {
  height: 20px;
  padding: 5px 10px;
  border-radius: 10px;
  border: 2px solid $default_color;
  font-size: 14px;
  font-weight: 400;
  color: $combobox_color;
}

.option-list {
  border: 1px solid $list_color;
  list-style: none;
  margin-top: 0;
  max-height: 143px;
  overflow-y: auto;
  padding-left: 0;
  width: calc(172px + 1rem);
  padding: 5px;
  border-radius: 10px;
  color: $combobox_color;
  border: 1px solid $option_list_color;
}

.option-list {
  &::-webkit-scrollbar {
    width: 12px;
  }

  &::-webkit-scrollbar-track {
    background: $scroll_color;
    border-radius: 10px;
  }

  &::-webkit-scrollbar-thumb {
    border-radius: 10px;
    border: 3px solid $default_color;
  }

}

.no-option,
.option {
  margin: 10px 5px;
  padding: 5px;

  &:hover:not(.no-option) {
    background: $default_color;
    color: $white_color;
    border-radius: 10px;
  }
}

