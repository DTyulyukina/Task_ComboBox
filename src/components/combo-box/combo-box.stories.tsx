import React, { useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { ComboBox } from './combo-box';

export default {
  title: 'ComboBox',
  component: ComboBox,
  argTypes: {
    value: {
      name: 'value',
      description: 'Ð¡omponent accepting value',
      required: true,
      control: {
        type: null
      },
      defaultValue: null
    },
    onChange: {
      name: 'onChange',
      description: 'Function change ComboBox',
      control: {
        type: 'Event'
      },
      required: true
    },
    optionList: {
      name: 'optionList',
      description: 'Accepts a list of options ComboBox',
      options: [
        'Item 1',
        'Item 2',
        'Item 3',
        'Item 4',
        'Item 5',
        'Item 6',
        'Item 7'
      ],
      control: {
        type: 'array'
      },
      required: true
    },
    defaultValue: {
      name: 'defaultValue',
      description: 'Default value could be specified',
      defaultValue: 'Item 2',
      control: {
        type: 'text'
      }
    }
  },
  parameters: {
    docs: {
      description: {
        component:
          'The _ComboBox_ component with a list of options (values for autocompletion)'
      }
    },
    actions: {
      handles: ['mouseleave', 'focus', 'keydown', 'click']
    }
  }
} as ComponentMeta<typeof ComboBox>;

const defaultComboBox: ComponentStory<typeof ComboBox> = ({
  optionList,
  defaultValue
}) => {
  const [value, setValue] = useState<string | undefined>();

  return (
    <ComboBox
      value={value}
      onChange={setValue}
      optionList={optionList}
      defaultValue={defaultValue}
    />
  );
};

export const comboBoxComponent = defaultComboBox.bind({});
comboBoxComponent.args = {
  defaultValue: 'Item 2',
  optionList: [
    'Item 1',
    'Item 2',
    'Item 3',
    'Item 4',
    'Item 5',
    'Item 6',
    'Item 7'
  ]
};
