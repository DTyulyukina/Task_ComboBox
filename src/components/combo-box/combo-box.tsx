import React, { createRef, useEffect, useMemo, useState } from 'react';

import './combo-box.scss';

interface ComboBoxProps {
  value: string | undefined;
  onChange: (value: string) => void;
  optionList: string[];
  defaultValue?: string;
}

enum ECodeKeys {
  ENTER = 'Enter'
}

export const ComboBox: React.FC<ComboBoxProps> = ({
  value,
  onChange,
  optionList,
  defaultValue
}) => {
  const [valueComboBox, setValueComboBox] = useState<string | undefined>('');
  const refComboBox = createRef<HTMLInputElement>();
  const [isShowListComboBox, setIsShowListComboBox] = useState<boolean>(false);

  useEffect(() => {
    defaultValue && value === undefined
      ? setValueComboBox(defaultValue)
      : setValueComboBox(value);
  });

  const optionListComboBox = useMemo(
    () =>
      value
        ? optionList.filter((option) =>
            option.toLowerCase().includes(value.toLowerCase())
          )
        : optionList,
    [value, optionList]
  );

  const addListValue = (
    event: React.SyntheticEvent<EventTarget>,
    index: number
  ) => {
    event.stopPropagation();

    onChange(optionListComboBox[index]);

    setIsShowListComboBox(false);
  };

  const changeValue = (event: React.SyntheticEvent<EventTarget>) => {
    onChange((event.target as HTMLInputElement)?.value);
  };

  const updateShowListOptions = () => {
    setIsShowListComboBox(true);

    refComboBox.current?.focus();
  };

  const updateHideListOptions = () => {
    setIsShowListComboBox(false);
  };

  const onKeyDownComboBox = (event: React.KeyboardEvent) => {
    if (event.key === ECodeKeys.ENTER) {
      refComboBox.current?.blur();

      onChange(refComboBox.current?.value as string);
    }
  };

  const valueOptions = useMemo(() => {
    return (
      <div
        className="option-list"
        style={{ display: isShowListComboBox ? 'block' : 'none' }}
      >
        {optionListComboBox?.length ? (
          optionListComboBox.map((item, index) => {
            return (
              <div
                className="option"
                key={item}
                onClick={(event) => addListValue(event, index)}
              >
                {item}
              </div>
            );
          })
        ) : (
          <div className="no-option">No option..</div>
        )}
      </div>
    );
  }, [optionListComboBox, value, isShowListComboBox]);

  return (
    <form
      style={{ display: 'inline-block', height: 'auto' }}
      onClick={updateShowListOptions}
      onFocus={updateShowListOptions}
      onMouseLeave={updateHideListOptions}
      onKeyDown={onKeyDownComboBox}
    >
      <input
        className="combo-box"
        onChange={changeValue}
        value={valueComboBox}
        ref={refComboBox}
      />
      {valueOptions}
    </form>
  );
};
