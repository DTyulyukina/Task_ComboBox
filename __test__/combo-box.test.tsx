import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';
import { ComboBox } from '../src/components/combo-box';

import '@testing-library/jest-dom';

const optionsTest = ['winter', 'spring', 'summer', 'autumn'];
const valueTest = '';
const changeTestMock = jest.fn();

describe('Combobox Component', () => {
  test('Renders correctly Combobox component', () => {
    const tree = create(
      <ComboBox
        value={valueTest}
        optionList={optionsTest}
        onChange={changeTestMock}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders form with add value in Combobox', async () => {
    render(
      <ComboBox
        value={valueTest}
        optionList={optionsTest}
        onChange={changeTestMock}
      />
    );

    fireEvent.change(screen.getByDisplayValue(''), {
      target: { value: '' }
    });

    await waitFor(() => {
      expect(screen.getByDisplayValue('')).toBeInTheDocument();
    });
  });
});
