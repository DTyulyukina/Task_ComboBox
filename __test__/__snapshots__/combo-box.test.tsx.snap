// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox Component Renders correctly Combobox component 1`] = `
<form
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "inline-block",
      "height": "auto",
    }
  }
>
  <input
    className="combo-box"
    onChange={[Function]}
    value=""
  />
  <div
    className="option-list"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="option"
      onClick={[Function]}
    >
      winter
    </div>
    <div
      className="option"
      onClick={[Function]}
    >
      spring
    </div>
    <div
      className="option"
      onClick={[Function]}
    >
      summer
    </div>
    <div
      className="option"
      onClick={[Function]}
    >
      autumn
    </div>
  </div>
</form>
`;
